packages/slate-react/src/hooks/use-stable.ts

```diff
@@ -0,0 +1,15 @@
+import { useRef } from 'react'
+import { ReactEditor } from '../plugin/react-editor'
+
+// Uninitialized value is an object so that no value a user could pass in can === it
+const UNINITIALIZED = {}
+// This helper function is required to help typescript track the type of ref.current below
+const isUninitialized = (v: unknown): v is typeof UNINITIALIZED =>
+  v === UNINITIALIZED
+
+export const useStable = <T>(init: () => T): T => {
+  const ref = useRef<T | typeof UNINITIALIZED>(UNINITIALIZED)
+  // If the ref hasn't been initialized, initialize it; else don't. then return the initialized value
+  // This is written as a single expression to help typescript understand what's going on.
+  return isUninitialized(ref.current) ? (ref.current = init()) : ref.current
+}
```

Comment:
I don't see why we have to use `UNINITIALIZED`.  Is it needed?  Can we just do:

```suggestion
export const useStable = <T extends Editor>(init: () => T): T => {
  const ref = useRef<T | null>(null)
  
  if (!ref.current) ref.current = init();
  
  return ref.current;
}
```

Reply:
For the slate use case it's effectively the same; but if a slate user also uses the `useStable` hook for other values, it's possible that their initialization function returns a falsey value, causing it to re-run each time the hook is called & breaking the contract of only running a single time. Add side effects & there's a hard to find bug.

Because this is a helper being exported to users to make maintaining the editor state reference easier, and not just an internal helper, it should uphold the contract, & not lead to hard to debug errors when used in other ways

Reply:
Ah I see what you are saying.  Do we need to expose a general hook?  The use case here is to ensure referential equality on the `editor`, no?  I might suggest limiting its use and calling it `useStableEditor` or the like as it's not really the province of Slate to expose a general hook.